sam.c

```
@@ -1948,33 +2569,98 @@ int sam_read1(htsFile *fp, sam_hdr_t *h, bam1_t *b)
+                return -1;
+            }
+
+            sp_bams *gb = fd->curr_bam;
+            if (!gb) {
+                hts_tpool_result *r = hts_tpool_next_result_wait(fd->q);
+                fd->curr_bam = gb = (sp_bams *)hts_tpool_result_data(r);
+                hts_tpool_delete_result(r, 0);
+            }
+            if (!gb)
+                return fd->errcode ? -2 : -1;
+            bam1_t *b_array = (bam1_t *)gb->bams;
+            if (fd->curr_idx < gb->nbams)
+                bam_copy1(b, &b_array[fd->curr_idx++]);
+            if (fd->curr_idx == gb->nbams) {
+                pthread_mutex_lock(&fd->lines_m);
+                gb->next = fd->bams;
+                fd->bams = gb;
+                pthread_mutex_unlock(&fd->lines_m);
+
+                fd->curr_bam = NULL;
+                fd->curr_idx = 0;
+            }
+
+            return 0;
+
+        } else  {
+            int ret;
+        err_recover:
+
             ret = hts_getline(fp, KS_SEP_LINE, &fp->line);
             if (ret < 0) return ret;
+
+            ret = sam_parse1(&fp->line, h, b);
+            fp->line.l = 0;
+            if (ret < 0) {
+                hts_log_warning("Parse error at line %lld", (long long)fp->lineno);
+                if (h->ignore_sam_err) goto err_recover;
+            }
+            return ret;
         }
-        ret = sam_parse1(&fp->line, h, b);
-        fp->line.l = 0;
-        if (ret < 0) {
-            hts_log_warning("Parse error at line %lld", (long long)fp->lineno);
-            if (h->ignore_sam_err) goto err_recover;
-        }
-        return ret;
     }
 
     default:
         abort();
     }
 }
 
-int sam_format1(const sam_hdr_t *h, const bam1_t *b, kstring_t *str)
+static int sam_format1_append(const bam_hdr_t *h, const bam1_t *b, kstring_t *str)
 {
     int i;
     uint8_t *s, *end;
     const bam1_core_t *c = &b->core;
 
-    str->l = 0;
+    if (c->l_qname == 0)
+        abort();
```

Better to return an error here.